File: server/app/flash.py
"""Active-flash liveness — the technique iProov, AWS Rekognition Face Liveness
and FaceTec all lean on, done with only the screen and camera we already have.

The screen flashes a **server-issued random colour sequence**; a real 3-D face
lit by the screen reflects each colour, so the mean colour of the face region
tracks the commanded sequence. A printed photo, a replayed video, or an injected
stream cannot follow a sequence it did not know in advance — so the face colour
does *not* track it. One technique that at once:

* is a challenge the attacker cannot pre-compute (the sequence is random),
* defeats replay (a recording of a *different* session's flashes won't correlate),
* defeats a flat photo / screen (its colour is fixed regardless of the flash),
* resists injection (a pre-rendered stream can't match realtime reflectance).

Pure and numeric on purpose: `flash_liveness_score` takes the commanded colours
and the observed mean face colours and returns 0..1, so the whole thing is
tested exhaustively on synthetic frames without a camera. Measuring the mean
face colour from real pixels is the caller's job (`services.verification`).
"""

from __future__ import annotations

import numpy as np

#: Full-screen flash colours, linear 0..1 RGB. Off-primaries (a channel is lit
#: to 1.0, the others to a floor) keep the screen bright enough to read a face
#: while still separating the channels — a pure (1,0,0) leaves the face almost
#: black in green/blue and the correlation there becomes noise.
FLASH_PALETTE: dict[str, tuple[float, float, float]] = {
    "red": (1.0, 0.15, 0.15),
    "green": (0.15, 1.0, 0.15),
    "blue": (0.15, 0.15, 1.0),
    "white": (1.0, 1.0, 1.0),
}

#: Below this score the face did not track the flash → treat as spoof/injection.
#: Real synthetic faces score ~0.85+; photo/replay/wrong-sequence spoofs score
#: near 0. 0.5 sits in the empty middle (see test_flash.py); tune on device.
FLASH_MIN = 0.5

#: A flash challenge needs at least this many frames for a stable correlation.
FLASH_MIN_FRAMES = 3


def flash_liveness_score(
    commanded: list[tuple[float, float, float]] | np.ndarray,
    observed: list[tuple[float, float, float]] | np.ndarray,
) -> float:
    """How well the face colour tracked the commanded flash sequence, 0..1.

    ``commanded`` — the colours the screen showed, one per flash frame (0..1 RGB).
    ``observed``  — the mean colour of the face region in each matching frame.

    Per channel, the Pearson correlation across frames between what the screen
    commanded and what the face showed. A real face is (roughly) an affine
    function of the illumination — ``observed ≈ ambient + albedo · commanded`` —
    so each channel the screen *varied* correlates near +1. Averaged over the
    channels the screen actually varied, negatives clamped to 0 (anti-correlation
    is not liveness). A dead-constant face (a photo) scores 0.

    Absolute brightness is deliberately ignored: ambient light and skin tone set
    the level, the *flash* sets the variation, and only the variation is trustworthy.
    """
    c = np.asarray(commanded, dtype=np.float64)
    o = np.asarray(observed, dtype=np.float64)
    if c.ndim != 2 or c.shape[1] != 3 or c.shape != o.shape:
        return 0.0
    if c.shape[0] < FLASH_MIN_FRAMES:
        return 0.0

    per_channel: list[float] = []
    for ch in range(3):
        commanded_ch = c[:, ch]
        observed_ch = o[:, ch]
        # The screen must have varied this channel for it to carry information.
        if commanded_ch.std() < 1e-3:
            continue
        # A face that does not var at all under a varying flash is a flat surface.
        if observed_ch.std() < 1e-6:
            per_channel.append(0.0)
            continue
        corr = float(np.corrcoef(commanded_ch, observed_ch)[0, 1])
        if np.isnan(corr):
            corr = 0.0
        per_channel.append(max(0.0, corr))

    if not per_channel:
        return 0.0
    return float(np.mean(per_channel))

File: packages/react-native-ekyc-local/src/flash.ts
/**
 * Active-flash liveness on the phone — a port of the server's `app/flash.py`.
 *
 * The screen flashes a random sequence of colours; a real 3-D face lit by the
 * screen reflects them, so the mean colour of the face tracks the commanded
 * sequence (per channel, Pearson correlation across frames). A printed photo
 * or a phone screen held up does not follow a sequence it never knew.
 *
 * Plainly: this catches **photos and screen replays**. It does **not** catch a
 * mask — a mask is a 3-D surface and reflects the screen much like skin. The
 * mask counter-measures are the movement challenges (a rigid mask cannot
 * open its mouth) and, in the heavy edition, rPPG.
 *
 * Pure maths here; sampling the face colour from frames is `colour.ts`.
 * Calibrated on synthetic reflectance (server tests) and, on 2026-08-19, four
 * real Railway sessions scored 0.82–0.92 with a real face — so 0.5 is a
 * comfortable gate; still shipped advisory in the local flow until its own
 * device numbers exist.
 */

export type Rgb = [number, number, number]

/** Same off-primaries as the server: bright enough to light a face, still separable per channel. */
export const FLASH_PALETTE: Record<string, Rgb> = {
  red: [1.0, 0.15, 0.15],
  green: [0.15, 1.0, 0.15],
  blue: [0.15, 0.15, 1.0],
  white: [1.0, 1.0, 1.0],
}
export const FLASH_MIN = 0.5
export const FLASH_MIN_FRAMES = 3

export function flashHex(name: string): string {
  const c = FLASH_PALETTE[name] ?? FLASH_PALETTE.white!
  return '#' + c.map((v) => Math.round(v * 255).toString(16).padStart(2, '0')).join('')
}

/** A random permutation of the palette (4 colours → every channel varies). */
export function pickFlashSequence(random: () => number = Math.random): string[] {
  const names = Object.keys(FLASH_PALETTE)
  for (let i = names.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1))
    ;[names[i], names[j]] = [names[j]!, names[i]!]
  }
  return names
}

function pearson(a: number[], b: number[]): number {
  const n = a.length
  const ma = a.reduce((s, v) => s + v, 0) / n
  const mb = b.reduce((s, v) => s + v, 0) / n
  let num = 0
  let da = 0
  let db = 0
  for (let i = 0; i < n; i++) {
    const x = a[i]! - ma
    const y = b[i]! - mb
    num += x * y
    da += x * x
    db += y * y
  }
  if (da <= 1e-18 || db <= 1e-18) return 0
  return num / Math.sqrt(da * db)
}

function std(v: number[]): number {
  const m = v.reduce((s, x) => s + x, 0) / v.length
  return Math.sqrt(v.reduce((s, x) => s + (x - m) ** 2, 0) / v.length)
}

/**
 * 0..1: how well the observed face colours tracked the commanded flash
 * colours. Per channel the screen varied, the correlation across frames,
 * negatives clamped to 0, averaged. A constant face (photo) scores 0.
 */
export function flashLivenessScore(commanded: Rgb[], observed: Rgb[]): number {
  if (commanded.length < FLASH_MIN_FRAMES || commanded.length !== observed.length) return 0
  const scores: number[] = []
  for (let ch = 0; ch < 3; ch++) {
    const c = commanded.map((v) => v[ch]!)
    const o = observed.map((v) => v[ch]!)
    if (std(c) < 1e-3) continue
    if (std(o) < 1e-6) {
      scores.push(0)
      continue
    }
    scores.push(Math.max(0, pearson(c, o)))
  }
  if (scores.length === 0) return 0
  return scores.reduce((s, v) => s + v, 0) / scores.length
}
